inspektor-gadget/inspektor-gadget
# Fix race condition on enrichers

Different gadgets use the different enricher interfaces to add
container information to the events received from the eBPF programs.
It could happen that by the time the tracer calls the enricher, the
container is gone and then the container information from the event
is missing.

This case is specially relevant when the container is killed after
executing the event, like using a capability not allowed or called
a syscall blocked by seccomp.

This commit fixes this by creating a container cache system: containers
are retained and made available to enrichers for a while after those
are removed from the container collection package, this allows the
enrichers to provide the container data even if the container is gone.

The container collection package is generic enough, hence this behavior
is only enabled when "WithTracerCollection" is used.


### How to test

Check https://github.com/inspektor-gadget/inspektor-gadget/issues/1178#issuecomment-1396238600. (Check that you're able to reproduce without this PR first, because it's possible that the issue doesn't reproduce in all environments)

### TODO
- [x] Implement test for WithTracerCollection
- [x] Update [audit seccomp](https://github.com/inspektor-gadget/inspektor-gadget/blob/00b892c97a95e0215cde92ec1e8c11d9349c0ee6/pkg/gadgets/audit/seccomp/tracer/tracer.go#L143-L151) to use this mechanism instead of saving container names in ebpf code? (cc @alban)

Fixes https://github.com/inspektor-gadget/inspektor-gadget/issues/1178
